### ISSUE TITLE
fonts for text - standard and incompatibility with (other) CAD programs

### ISSUE THREAD

**Issue by dec1** [mozman/ezdxf#325]
Hello.

I can read/write dxf files containing text, which have fonts (styles) specified, using eg [this](https://github.com/mozman/ezdxf/issues/243#issuecomment-713598895).

However, Ive tried several (widely used) CAD applications and none of them seem to be able to read the font (names) correctly in the files that ezdxf writes (and similarly ezdxf doesnt read the font (names) correctly in dxf files created by such such a CAD programs). 

Ive been told that the dxf standard is a bit unclear on how exactly to write the font info into. a dxf file.
Would you agree, or would you say the standard specifies  a single clear and unambiguous way to specify in the dxf file what font (if any) a piece of text has, and that ezdxf adheres to  this completely?

Best Wishes.
Declan

**Comment by mozman** [comment]
This is a [link](http://help.autodesk.com/view/OARX/2018/ENU/?guid=GUID-EF68AF7C-13EF-45A1-8175-ED6CE66C8FC9) to the DXF reference, which shows the text style (fonts) definition, that's all.

**Comment by dec1** [comment]
Thank you. 

Is it unambiguous what "style name" (2) here means.
This is perhaps what some CD programs show in the GUI as the "font name" and maybe they have some freedom to call this what they want?

The a  link above is a (defacto) standard?

Thanks.

**Comment by mozman** [comment]
There is no "DXF Standard", there is only this DXF reference provided by Autodesk, everything which is not described by the reference (and that's a lot), is defined by the results created by AutoCAD and has to be reverse engineered by the implementer.

**Comment by dec1** [comment]
So if another application isnt reading/writing dxf files in a way compatible with AutoCAD one could  venture to say they  are doing something "wrong"?

And AutoCAD itself is presumably also conforming to the [autodesk reference](http://help.autodesk.com/view/OARX/2018/ENU/?guid=GUID-EF68AF7C-13EF-45A1-8175-ED6CE66C8FC9)?

**Comment by mozman** [comment]
If there is only a shallow DXF reference, what do you mean by "wrong". The result may not look like in AutoCAD but could be conform to all documented DXF features - **the DXF reference provides no rules how to render entities**. There exist not a single CAD application which renders DXF or DWG files to 100% like AutoCAD. 

I often try to explain to people who demand more rendern capabilities from ezdxf, that the DXF format is a data exchange format in the first place and not a layout preserving exchange format like PDF. Wikipedia: The PDF file encapsulates a complete description of a fixed-layout flat document, including the text, fonts, vector graphics, raster images. All this layout rules do not exist  in DXF and the biggest difference: there is no [ISO Standard](https://www.iso.org/standard/51502.html) for DXF

**Comment by dec1** [comment]
> what do you mean by "wrong"

For example,  I create some text in a CAD application, apply a common font with a certain name, and see the appearance of the text changes. If I then export this to a dxf file and then re-import it I would expect the text to look the same (ie font file was found and applied as intended) , and the ui (text properties dialog or whatever) to show the same font name as before the export. 

If I instead import the same file into a different application (that also "supports" this font) on the same machine/os (ie same system fonts installed), and I see the text has a default appearance (ie as when no font applied)  and the font settings dialog to display "system" or "default", instead of something to suggest the original name then Id say:

- The applications are not compatible as far exporting text with fonts are concerned, and 
- Either  there isnt any standard reference that describes unambiguously how they should both be reading/writing the relevant sections of the dxf file, or one or both are not adhering to this reference/standard (ie is "wrong")

If one of the applications is AutoCAD and the dxf files it writes can be regarded as a de-facto reference for anything not specified enough in  the [reference](http://help.autodesk.com/view/OARX/2018/ENU/?guid=GUID-EF68AF7C-13EF-45A1-8175-ED6CE66C8FC9), then Id say its the other application is "wrong"

**Comment by dec1** [comment]
Are you aware of any semantic description of whats meant by "Style name" [here](http://help.autodesk.com/view/OARX/2018/ENU/?guid=GUID-EF68AF7C-13EF-45A1-8175-ED6CE66C8FC9) anywhere else in the documentation?

**Comment by mozman** [comment]
Text style is the definition of font to use for each text entity which references this text style by name (group code 2).

**Comment by dec1** [comment]
Let me maybe rephrase my question.

If I:

- create a drawing with AutoCAD, that just has a single piece of text with a certain font applied
- export this to file1.dxf from AutoCAD
- import file1.dxf this using ezdxf
- export a new file2.dxf based on the info ezdxf gathered in the previous import
- import file2.dxf (into a new drawing) in AutoCAD (this and the next step just uses code 

Would you expect drawing 1 and drawing 2 to look the same in AutoCAD (ie same text and same text font applied)?
And that the python code necessary to do this would be [here](https://github.com/mozman/ezdxf/issues/243#issuecomment-713598895)? In particular the values "my_str" and "font"


Thanks in Advance

**Comment by mozman** [comment]
Text created with font BOUTON in BricsCAD, the border should show if the text size is changing:

![BOUTON-BricsCAD](https://user-images.githubusercontent.com/4230966/105842655-d0768380-5fd6-11eb-8048-065336bd230c.jpg)

File loaded and saved by ezdxf - displayed in BricsCAD:

![BOUTON-ezdxf-BricsCAD](https://user-images.githubusercontent.com/4230966/105842728-ef751580-5fd6-11eb-98ea-c4227e07438d.jpg)

File loaded and saved by ezdxf - displayed in TrueView (Autodesk viewer):
![BOUTON-ezdxf-TrueView](https://user-images.githubusercontent.com/4230966/105842778-061b6c80-5fd7-11eb-9b48-54dd25d906a5.jpg)

DXF Files: [ezdxf.zip](https://github.com/mozman/ezdxf/files/5873155/ezdxf.zip)
- bouton.dxf - created in BricsCAD
- bouton-ezdxf.dxf - loaded and saved by ezdxf

**Comment by dec1** [comment]
Thank you for the examples. Can you please also post the python code you used to read and write these?

**Comment by mozman** [comment]
```Python
import ezdxf

doc = ezdxf.readfile("bouton.dxf")
doc.saveas("bouton-ezdxf.dxf")
```

**Comment by dec1** [comment]
Thank you. That indeed works for me too in tests with my cad program (qcad). However I need to be able to write not just the exact same content but, optionally edit the document and (manually) save the new version. 

For example:

I have a  dxf file (in.dxf - see attached zip archive) created by autoCAD. I can view it in my cad viewer (qcad) and everything is as expected. I see the correct text content ("Abcd123"), the text properties dialog  shows "Arial", and the text is rendered as I would expect (in particular its not in the default "sytle").

(If I just rewrite this as a dxf file exactly as is using your simple code above, then this new file would also show as expected in my cad viewer).

 However, if I read this with python/ezdxf, and then write another dxf file (See attached out.dxf) using code as below, things are not as I would expect. If I open this out.dxf in my cad program, the contents of the text is fine, but the text properties dialog shows "arial.ttf", and the text is rendered in the default style. I can instead choose "Arial" from the text properties drop down menu and the text then changes appearance and looks like "in.dxf".

I think I must be doing something wrong (or omitting to do something right) when writing the dxf file. Maybe there are some attributes Im omitting to read and then set/write. Can you please tell me if you can see whats amiss?

Many Thanks in Advance.

```import ezdxf

`

def write_it(my_str_font, file_path_out):
    doc_out = ezdxf.new('R2018')
    msp_out = doc_out.modelspace()

    for my_str, my_font in my_str_font:

        # create text "styles" (map names to font files)
        doc_out.styles.  new('myStyle', dxfattribs={'font': my_font})

        # Create a text item with a text style
        my_text = msp_out.add_mtext(my_str, dxfattribs={'style':'myStyle'})

    doc_out.saveas(file_path_out)

# --------------------------------------------------------------------------------------------
def read_it(file_path):
    doc_in = ezdxf.readfile(file_path)

    styles_in = doc_in.styles
    msp_in = doc_in.modelspace()

    my_str_font = []

    for e in msp_in:

        my_str = None
        my_font = None
        Type = e.dxftype()
        if (Type == "TEXT"):
            my_str = e.get_dxf_attrib("text")
            style_name = e.get_dxf_attrib("style")
            style = styles_in.get(style_name)
            my_font = style.dxf.font
            halt =1
        elif (Type == "MTEXT")    :
            my_str = e.text
            style_name = e.get_dxf_attrib("style")
            style = styles_in.get(style_name)
            my_font = style.dxf.font
            halt =1

        if(my_str):
            my_str_font.append((my_str, my_font))

    return my_str_font

# --------------------------------------------------------------------------------------------
file_path_in = "<...replace_with_local_path_to...>/in.dxf"
file_path_out = "<...replace_with_local_path_to...>/out.dxf"

ret = read_it(file_path_in)

write_it(ret, file_path_out)
```



[autcad_in_ezdxf_out.zip](https://github.com/mozman/ezdxf/files/5879328/autcad_in_ezdxf_out.zip)

**Comment by mozman** [comment]
I don't understand what you exactly want, but if the problem is that QCAD does not show same result as AutoCAD, then it is a problem of QCAD.

BTW: LibreCAD and QCAD are not reliable tools for rendering DXF content. I do not accept bug reports, which are based on rendering issues of these two "CAD" applications.

**Comment by dec1** [comment]
Its nothing to do with any particular CAD application as far as I can see.

Whats missing it seems, is a call to `style.set_extended_font_data` as below

```    
  my_text = msp_out.add_mtext("hello", dxfattribs={'style':'myStyle'})
  style.set_extended_font_data("Arial")
```
This feature has only been [recently](https://ezdxf.mozman.at/release-v0-15.html) been added to ezdxf, it seems. Maybe it might be worth adding it to the documentation.

Your snippet 
```
doc = ezdxf.readfile("bouton.dxf")
doc.saveas("bouton-ezdxf.dxf")
```
(manually) does the same implicitly somehow

**Comment by mozman** [comment]
I don't know if this changes anything. AutoCAD and BricsCAD work without extended data and other CAD applications do not care about this extended data.

**Comment by dec1** [comment]
It seems like AutoCad does use this extended font data. It certainly writes it to dxf files.
In the zip file (autcad_in_ezdxf_out.zip) I attached above is a dxf file exported from autoCAD (in.dxf).
Reading it with code from above but modified to additionally check the extended data:

```
# --------------------------------------------------------------------------------------------
def read_it(file_path):
    ....< snip (see above)> 
           style_name = e.get_dxf_attrib("style")
            style = styles_in.get(style_name)
            my_font = style.dxf.font

            has_ext_data = style. has_extended_font_data
            font_ext_data = style.get_extended_font_data()
            Family = font_ext_data[0]

      .... <snip ..> 
```

produces `has_ext_data = True,` and `Family = "Arial" `
